035-cruise-control-scaling-brokers.md

```diff
@@ -0,0 +1,113 @@
+# Cruise Control rebalancing on scaling the Apache Kafka cluster
+
+This proposal is about leveraging the Cruise Control integration in the Strimzi cluster operator in order to run an Apache Kafka cluster rebalancing when it's scaled up or down.
+
+## Current situation
+
+Currently, when Cruise Control is deployed alongside the Apache Kafka cluster, no automatic rebalancing happens when the cluster itself is scaled up or down.
+
+It means that, if the cluster is scaled up, no replicas will be moved to the newly added brokers.
+These brokers will only get partition replicas of newly created topics.
+Furthermore, by using the `KafkaRebalance` custom resource to run a rebalancing operation after the scaling up, it will move replicas across all the already existing brokers and not only to the new ones.
+It actually runs a full-rebalance with the drawback of more data movement, CPU usage and so on.
+
+In the same way, if the cluster is scaled down, the replicas hosted on the target brokers to remove are not moved out before the scaling down happens.
+Right now the scaling down is not blocked but doing it is not part of this proposal.
+This could result in some partitions having less ISR (In-Sync Replicas) than before the scaling down, with an impact on availability of the cluster.
+The current Cruise Control integration has no support, via the `KafkaRebalance` custom resource, to allow moving these replicas out of the target brokers before the scaling down happens.
+
+## Motivation
+
+When the Apache Kafka cluster is scaled up, it should be possible to use the Cruise Control integration to run a rebalancing by moving some replicas to the newly added brokers without involving the already existing ones.
+This would provide a faster rebalancing, not involving the entire cluster.
+
+Not stopping the scaling down operation when replicas are hosted on the target brokers to remove is not part of this proposal but it should be possible to move replicas out of the target brokers before the scaling down happens.
+It would allow to remove "empty" brokers which are not hosting any replicas, avoiding less ISR problem.
```

Comment:
I find this point a bit weird. Not stoping the scaledown when the brokers are not empty can be seen as a _missing feature_ today. But how does it fit with this proposal? Is this something to consider as a feature when the operator can move the replicas out of the broker automatically?

Reply:
> Not stopping the scaling down operation when replicas are hosted on the target brokers to remove is not part of this proposal ....

I was trying to make it clear that this proposal is about moving out replicas from brokers and then scaling them down. There is no operator feature that stops a user to scale down when a broker is hosting replicas. AFAIU it's something that @ShubhamRwt is looking into by using Admin API. If you feel that this sentence just creates confusion, we can remove it.